ratik/src/intTest/java/org/opencadc/ratik/InventoryValidatorTest.java

```diff
@@ -0,0 +1,855 @@
+            System.setProperty("user.home", USER_HOME);
+        }
+
+        Artifact localArtifact = this.localEnvironment.artifactDAO.get(artifact.getID());
+        Assert.assertNotNull("local artifact not found", localArtifact);
+        Assert.assertEquals("metaChecksum mismatch", artifact.getMetaChecksum(), localArtifact.getMetaChecksum());
+        Assert.assertTrue("artifact does not contains remote site location",
+                          localArtifact.siteLocations.contains(new SiteLocation(remoteSiteID)));
+        Assert.assertTrue("artifact does not contains local site location",
+                          localArtifact.siteLocations.contains(new SiteLocation(randomSiteID)));
+    }
+
+    /**
+     * explanation6: deleted from L, lost DeletedArtifactEvent
+     * evidence: ?
+     * action: assume explanation3
+     *
+     * explanation7: L==storage, deleted from L, lost DeletedStorageLocationEvent
+     * evidence: ?
+     * action: assume explanation3
+     *
+     * explanations 6 and 7 are covered by explanation 3
+     */
+
+    /**
+     * discrepancy: artifact.uri in both && artifact.id mismatch
+     *
+     * explanation1: same ID collision due to race condition that metadata-sync has to handle
+     * evidence: no more evidence needed
+     * action: pick winner, create DeletedArtifactEvent for loser, delete loser if it is in L
+     *
+     * case 1:
+     * local Artifact contentModifiedDate before remote Artifact contentModifiedDate, delete local Artifact.
+     * before: Artifact.uri in L && R, L Artifact.id != R Artifact.id,
+     *         L Artifact.contentModifiedDate < R Artifact.contentModifiedDate
+     * after:  Artifact not in L, DeletedArtifactEvent in L
+     *
+     * case 2:
+     * local Artifact contentModifiedDate after remote Artifact contentModifiedDate, no action
+     * before: Artifact.uri in L && R, L Artifact.id != R Artifact.id,
+     *         L Artifact.contentModifiedDate >= R Artifact.contentModifiedDate
+     * after:  Artifact in L
+     */
+    @Test
+    public void artifactUriCollision_LocalIsStorage() throws Exception {
+        artifactUriCollision(false);
+    }
+
+    @Test
+    public void artifactUriCollision_LocalIsGlobal() throws Exception {
+        artifactUriCollision(true);
+    }
+
+    public void artifactUriCollision(boolean trackSiteLocations) throws Exception {
+        // case 1: local contentModifiedDate before remote contentModifiedDate
+        URI contentCheckSum = TestUtil.getRandomMD5();
+        UUID localID = UUID.randomUUID();
+        UUID remoteID = UUID.randomUUID();
+        Calendar contentModifiedDate = Calendar.getInstance();
+
+        Artifact remoteArtifact = new Artifact(remoteID, URI.create("cadc:INTTEST/one.ext"), contentCheckSum,
+                                               contentModifiedDate.getTime(), 1024L);
+        this.remoteEnvironment.artifactDAO.put(remoteArtifact);
+
+        contentModifiedDate.roll(Calendar.YEAR, false);
```

Comment:
Since it is critical that the Artifact.uri values are equal, I would use remoteArtifact.getURI() in the localArtifact... or maybe a variable that both ctors use. Otherwise, everone has to read that string char-by-char and that's no fun :-)

Reply:
Cleaned up so you only have to read that string once!